src/Data/Fin/Relation/Unary/Top.agda

```diff
@@ -0,0 +1,84 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- The 'top' view of Fin
+--
+-- This is an example of a view of (elements of) a datatype,
+-- here i : Fin (suc n), which exhibits every such i as
+-- * either, i = fromℕ n
+-- * or, i = inject₁ j for a unique j : Fin n
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Data.Fin.Relation.Unary.Top where
+
+open import Data.Nat.Base using (ℕ; zero; suc; _<_)
+open import Data.Fin.Base using (Fin; zero; suc; toℕ; fromℕ; inject₁)
+open import Data.Fin.Properties using (toℕ-fromℕ; toℕ-inject₁; inject₁ℕ<)
+open import Relation.Binary.PropositionalEquality.Core
+open import Relation.Nullary.Negation using (contradiction)
+
+private
+  variable
+    n : ℕ
+    i : Fin (suc n)
+    j : Fin n
+
+------------------------------------------------------------------------
+-- The View, considered as a unary relation on Fin (suc n)
+
+-- NB `Data.Fin.Properties.fromℕ≢inject₁` establishes that the following
+-- inductively defined family on `Fin (suc n)` has constructors which
+-- target *disjoint* instances of the family; and hence the interpretations
+-- of the View constructors will also be disjoint
+
+data View : ∀ {n} (j : Fin n) → Set where
+  ‵fromℕ : View (fromℕ n)
+  ‵inj₁ : {j : Fin n} → View j → View (inject₁ j)
+
+pattern ‵inject₁ {n} j = ‵inj₁ {n = n} {j = j} _
+
+-- The view covering function, witnessing soundness of the view
+
+view-zero : ∀ n → View {suc n} zero
+view-zero zero    = ‵fromℕ
+view-zero (suc n) = ‵inj₁ (view-zero n)
+
+view : (j : Fin n) → View j
+view zero = view-zero _
+view (suc i) with view i
+... | ‵fromℕ  = ‵fromℕ
+... | ‵inject₁ j = ‵inj₁ (view (suc j))
+
+-- Interpretation of the view constructors
+
+⟦_⟧ : View {n} j → Fin n
```

Comment:
I'm a bit worried about how often the `{n}` argument in `View` appears. It seems to be needed in every single proof below, which suggests to me that it should be an explicit argument instead?

Reply:
So... I think we have been swirling around this issue for a while, and I very much appreciate you continuing to hold my feet to the fire over it ;-)

What I (and I think, you, as well ;) would like is for the `variable` declarations to do all the heavy lifting for me in terms of (telescopic, but that's not how it works in Agda... :-() typing and scoping of both `j` *and* `n`, so that when I mention `j` in the type of a definition, I get the `n` index in its type 'for free', and in particular, its *actual name/identifier* 'n'... but that's not how variable declaration generalisation works (see the [recent discussion on the Agda Zulip](https://agda.zulipchat.com/#narrow/stream/238741-general/topic/Referring.20to.20variables.20used.20by.20variables.3F) about `Set _` level inference... ).

So the 'correct' solution is probably to bite the bullet and write
```
⟦_⟧ : {j : Fin n} → View j → Fin n
```
which my pig-headedness otherwise prevents me from doing, in favour of *mentioning* the (correctly implicitly-bound: when considering `j`, the typechecker will never *not* know its index) name 'n' in order to be able to refer to it in the return type...

Alternatively (but even more opaquely!), to write
```
⟦_⟧ : View j → Fin _
```
(NB I need to check whether this works).

So what you see, and perhaps rightly complain about, is my tricksy-but-hamfisted attempt to workaround/futile protest against what I see as the 'problem' with Agda's approach to `varaiable` generalisation.

Reply:
Hopefully my most recent commits have now resolved this discussion?